### PR TITLE
release: 0.20.32 — ABI v10, EMBED_IMAGE, and NeuRT's embedding/rerank/VLM/vision slots

### PR DESCRIPTION
**Pull request by sanchitmonga22**
NeuRT filled 3 of 10 vtable slots, so bundles for embedding, rerank, VLM and
vision loaded and then had nothing to drive them. neurun
[#86](https://github.com/RunanywhereAI/neurun/pull/86) built the drivers; this is
the SDK half, plus the ABI bump they need and a public entry point for the one
primitive that had none.

## ABI 9 → 10

`RAC_PRIMITIVE_RESERVED_12` → `RAC_PRIMITIVE_EMBED_IMAGE` (wire 12) and
`reserved_slot_3` → `image_embedding_ops` — the same promote-in-place move rerank
made in v8. **No existing field moved**: struct size 232 and every offset are
identical to `main`, and `test_engine_vtable`'s 17-pointer tail assertion still holds.

## Image embedding is now callable, not just routable

The slot alone was not a feature — nothing could reach it. `rac_image_embedding_service.cpp`
is a standalone service factory in the shape rerank uses (no `LoadedModel` slot, no
`ModelCategory` arm, no component enum), plus a Swift bridge and
`RunAnywhere.embeddings.embedImage(_:)`.

## CoreML routing fails closed

`COREML` now maps to `neurt` unconditionally (the old per-primitive allow-list went
stale twice), but the priority fallback under it was a regression waiting to happen:
MLX is priority 110 to NeuRT's 100 and also fills `embedding_ops`, and `platform`
really does serve TTS. A Core ML model NeuRT cannot serve would have been handed to an
engine that cannot read an `.mlmodelc` at all. Now `BACKEND_NOT_FOUND`.

## `input_type` finally reaches the engine

`idl/embeddings_options.proto` has declared `EmbeddingsInputType` since it was written
and nothing read it. Not cosmetic: on Nemotron-3-Embed an unprefixed query/passage pair
separates a relevant passage from an irrelevant one by ~0.001; correctly prefixed, 0.48.
Surfaced as `EmbedOptions.inputType` and RCLI `embed --input-type`.

## Also

- The CRACommons mirror had already drifted — it never got the ABI-9 rerank promotion,
  so it still called wire 11 "reserved". The four ABI headers are now forwarders, which
  makes that class of drift impossible.
- Diffusion registry: both CoreML rows advertised DPM++ 2M against a DDIM-only engine,
  img2img against a text-to-image-only engine, and a `coreml_path` that does not exist
  in Apple's repo.
- `check_engine_prebuilt_pins.sh` looked for QHexRT assets in the **NeuRT** release
  listing. Latent until now; v0.20.32 is the first release to diverge the two tags.

## Verified against the published release, not a local build

```
download-neurt.sh --all        3/3 slices, SHA-256 verified, ABI 10
check_engine_prebuilt_pins.sh  online: pins, consumers and contents all agree
ctest                          111/111
vision  e2e   dim=768, distinct vectors per image (proves pixels reach the tower)
STT     e2e   WER 0.0000, 76 ms, coverage 100%
embedding     dim 2048, cat/kitten 0.4763 vs cat/earnings 0.0071
rerank        Paris 1.316 | Eiffel −5.676 | photosynthesis −7.188
diffusion     512×512 RGBA
swift build   clean · RCLI BUILD SUCCEEDED
```

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01R6p1R4pwTg5pvVjiEjq9ZA

## Summary by CodeRabbit

- **New Features**
  - Added image embedding support to the Swift SDK, converting images into vector embeddings.
  - Added query and document input types for text embedding options.
  - Expanded NeuRT support for embeddings, reranking, visual-language features, and image embeddings.

- **Bug Fixes**
  - Improved Core ML model routing with fail-closed behavior when the appropriate backend is unavailable.
  - Corrected built-in diffusion model configuration and capability reporting.

- **Release**
  - Released version 0.20.32 across supported SDKs and platforms.

### CODE REVIEW

File: core/tests/test_image_embedding_neurt_e2e.cpp [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+}  // namespace
+
+int main() {
+    std::fprintf(stdout, "test_image_embedding_neurt_e2e\n");
+
+    const char* bundle = std::getenv("RAC_TEST_NEURT_IMAGE_EMBED_BUNDLE");
+    if (!bundle || bundle[0] == '\0') {
+        std::fprintf(stdout,
+                     "SKIP: set RAC_TEST_NEURT_IMAGE_EMBED_BUNDLE to a SigLIP2 _ANE bundle\n");
+        return 0;
```
[the remark this answers is not in the document]
Comment (sanchitmonga22) [reply core/tests/test_image_embedding_neurt_e2e.cpp]: Fixed in `7b46abd28` — and this was the most valuable of the four.

You are right that the registration set no `SKIP_RETURN_CODE`, so the SKIP path was counted as a pass. The test now exits **77** and `CMakeLists.txt` registers `SKIP_RETURN_CODE 77`. Verified both directions:

```
no bundle : 1/1 Test #12: image_embedding_neurt_e2e_tests ...***Skipped   0.57 sec
bundle    : 1/1 Test #12: image_embedding_neurt_e2e_tests ...   Passed    3.59 sec
```

Worth recording why this matters here specifically: `test_stt_neurt_e2e` returns 0 on the same condition, and it has been reported green for its entire life without ever running. When it was finally given a real bundle during this work it failed immediately — every published ASR bundle is missing the `decode`/`audio` manifest blocks the runtime requires. A green tick that means "no bundle" is worse than a red one.

I deliberately did **not** change `test_stt_neurt_e2e` in this PR: it is not this change's file, and flipping it to 77 would turn `main` red for a reason unrelated to the ABI bump. It needs its bundles republished first.

File: core/tests/test_plugin_entry_neurt.cpp [hunk 2]
```diff
@@ -156,9 +156,36 @@ int main() {
...
+        if (vt->image_embedding_ops == nullptr) {
+            std::fprintf(stderr, "routable neurt engine has NULL image_embedding_ops slot\n");
+            return 1;
+        }
+        // Disjoint-slot invariant: DIFFUSION + GENERATE_TEXT + TRANSCRIBE + EMBED + RERANK + VLM +
+        // EMBED_IMAGE, and nothing else.
+        // Each name here is a capability PROMISE — the SDK treats a non-null pointer as "this engine
+        // serves that primitive" — so a slot must stay null until its driver is gated, not be filled
+        // speculatively.
+        if (vt->tts_ops != nullptr || vt->vad_ops != nullptr) {
```
[the remark this answers is not in the document]
Comment (sanchitmonga22) [reply core/tests/test_plugin_entry_neurt.cpp:188]: Fixed in `7b46abd28`. Confirmed against the code before changing it — the condition tested seven slots and omitted exactly the two you named:

```c
vt->diffusion_ops || vt->llm_ops || vt->stt_ops || vt->tts_ops ||
vt->vad_ops || vt->vlm_ops || vt->embedding_ops
```

Both `rerank_ops` (ABI v8) and `image_embedding_ops` (ABI v10, this PR) were promoted after that line was last touched, so a stub build advertising either would have passed a check whose entire purpose is "advertises nothing". Both added, with a comment naming the drift so the next promotion has a reason to look here — this is the same failure mode that left the CRACommons header mirror a whole ABI version behind.

File: bindings/flutter/packages/runanywhere_mlx/CHANGELOG.md [hunk 3]
```diff
@@ -1,3 +1,13 @@
+## [0.20.32] - 2026-08-31
+
+### Changed
+
+- No API change. Rebuilt against plugin ABI v10.
+- Core ML models are now routed to the NeuRT engine unconditionally and fail
+  closed when it cannot serve the primitive. Previously a Core ML model MLX
+  could not read could still be handed to MLX by plugin priority (110 vs
+  NeuRT's 100).
```
[the remark this answers is not in the document]
Comment (sanchitmonga22) [reply bindings/flutter/packages/runanywhere_mlx/CHANGELOG.md:9]: Fixed in `7b46abd28` — reworded to "a Core ML model **that** MLX could not read could still be handed to MLX by plugin priority (110 vs NeuRT's 100)". The missing relative pronoun genuinely made it parse two ways.

File: .agents/skills/source-command-copy-over-logic/SKILL.md [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+## 5. Quality + correctness checks - Launch a verification task
+
+- Ensure the **to** project still compiles.
+  - Fix imports, types, build settings, and module wiring.
+- Run any available tests relevant to the change.
+- If there were tests on the **from** side:
+  - Re-use/adapt them for the **to** side where practical, or at least mirror the key cases.
+- Manually sanity-check:
+  - Main happy path
+  - Important edge cases (failures, timeouts, offline scenarios, permission denials, etc.)
```
[the remark this answers is not in the document]
Comment (sanchitmonga22) [reply .agents/skills/source-command-copy-over-logic/SKILL.md]: Resolved differently in `7b46abd28`: **the file has been removed from this PR.**

`.agents/skills/source-command-copy-over-logic/SKILL.md` was untracked before this work began and a `git add -A` swept it into the ABI commit. That was my mistake, not a deliberate inclusion — an ABI promotion has no business carrying someone else's in-progress skill doc. `git rm --cached`, left on disk.

On the substance: your point about requiring executable evidence rather than "it compiles" is a good one, and it matches how this repo already works elsewhere — `sdk-test-starters` sets exactly that bar. But it belongs to whoever is authoring that skill, not to this change. I have not opened an issue for it, per the author's instruction.
